packages/utils/src/Signers.ts

```diff
@@ -0,0 +1,425 @@
+ * @returns A signer interface capable of making ES256K signatures.
+ */
+export async function es256kSigner({
+  secretKey,
+  keyUri,
+}: {
+  keyUri: string
+  secretKey: Uint8Array
+  publicKey?: Uint8Array
+}): Promise<SignerInterface> {
+  // only exists to map secretKey to seed
+  return es256kSignerWrapped({ seed: secretKey, keyUri })
+}
+
+/**
+ * Signer that produces an Ed25519 signature over the message.
+ *
+ * @param input Holds all function arguments.
+ * @param input.keyUri Sets the signer's id property.
+ * @param input.secretKey A 32 byte Ed25519 secret key. Some key representations append the public key to the private key; to allow these, all bytes after the 32nd byte will be dropped.
+ * @param input.publicKey The corresponding public key. May be omitted.
+ * @returns A signer interface capable of making Ed25519 signatures.
+ */
+export async function ed25519Signer({
+  secretKey,
+  keyUri,
+}: {
+  keyUri: string
+  secretKey: Uint8Array
+  publicKey?: Uint8Array
+}): Promise<SignerInterface> {
+  // polkadot ed25519 private keys are a concatenation of private and public key for some reason
+  return ed25519SignerWrapped({ seed: secretKey.slice(0, 32), keyUri })
+}
+
+/**
+ * Signer that produces an Sr25519 signature over the message.
+ *
+ * @param input Holds all function arguments.
+ * @param input.keyUri Sets the signer's id property.
+ * @param input.secretKey A 64 byte Sr25519 secret key.
+ * @param input.publicKey The corresponding 32 byte public key.
+ * @returns A signer interface capable of making Sr25519 signatures.
+ */
+export async function sr25519Signer({
+  secretKey,
+  keyUri,
+  publicKey,
+}: {
+  publicKey: Uint8Array
+  secretKey: Uint8Array
+  keyUri: string
+}): Promise<SignerInterface> {
+  await cryptoWaitReady()
+  return {
+    id: keyUri,
+    algorithm: ALGORITHMS.SR25519,
+    sign: async ({ data }: { data: Uint8Array }) => {
+      return sr25519Sign(data, { secretKey, publicKey })
+    },
+  }
+}
+
+function extractPk(pair: KeyringPair): Uint8Array {
+  const pw = randomAsHex()
+  const encoded = pair.encodePkcs8(pw)
+  const { secretKey } = decodePair(pw, encoded)
+  return secretKey
+}
+
+const signerFactory = {
+  [ALGORITHMS.ED25519]: ed25519Signer,
+  [ALGORITHMS.SR25519]: sr25519Signer,
+  [ALGORITHMS.ES256K]: es256kSigner,
+  [ALGORITHMS.ECRECOVER_SECP256K1_BLAKE2B]: polkadotEcdsaSigner,
+  [ALGORITHMS.ECRECOVER_SECP256K1_KECCAK]: ethereumEcdsaSigner,
+}
+
+/**
+ * Creates a signer interface based on an existing keypair and an algorithm descriptor.
+ *
+ * @param input Holds all function arguments.
+ * @param input.keyUri Sets the signer's id property.
+ * @param input.keypair A polkadot {@link KeyringPair} or combination of `secretKey` & `publicKey`.
+ * @param input.algorithm An algorithm identifier from the {@link ALGORITHMS} map.
+ * @returns A signer interface.
+ */
+export async function signerFromKeypair({
+  keyUri,
```

Comment:
The issue is knowing the KeyId vs KeyUri. It is one of the reasons why we moved away from having different terms.  I am not sure we do this now. We had problems in the past and that is why we have v1 & v3 for sporran.

Let us always take a keyUri and strip the unnecessary information magically. Hidden from those who don't want to go into the specifications :)

Reply:
Yeah I'm not a fan either, it's what we have introduced in the jcs repo. I'd either call it `keyId` like you suggested, or just plain `id`, because that is the property on the signer that it will be written to. Mind that this can also be an address in case this is not a Did-signer but an Account-signer. May be an argument for `id`, what do you think?

Reply:
@Dudleyneedham : this is a new interface, so won't be breaking old ones. But I did have the same hunch here: https://github.com/KILTprotocol/sdk-js/pull/799#discussion_r1369717461

Reply:
I have not answered to that issue yet, but I think that while it is fine to use `verificationMethod` for both ID and full node in jsonld, for the SDK this should be different, especially since we have a `VerificationMethod` type that includes the whole object. The use of `key` is not the right one, I feel. So my best choice would be `verificationMethodId`, which refers exactly to [this](https://www.w3.org/TR/did-core/#dfn-verificationmethod) property here.

Reply:
I think we should bring this to a meeting to discuss this point. It is important to see what we want as a team.

Reply:
Do we really need a meeting to discuss this? Can you bring your point forward? Honeslty except for you, me, Raphael, and Luke, and don't see anyone else that could provide an informed decision. Maybe Tino?

Reply:
I think I'll opt for just calling the parameter `id`, in line with the property it is written to, and explaining it's meaning in a docstring.

Reply:
f0e6c7496b8ad78e05167c9debd1e2595361a6a3